<a href="https://colab.research.google.com/github/joshlensing/Food-Sales-Predictions/blob/main/Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 1- Part 1
- Joshua Lensing

## Loading Data

In [87]:
import pandas as pd

In [88]:
filename = '/content/sales_predictions.csv'
df_sales = pd.read_csv(filename)
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [89]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Based on the first look of the data, it appears that there are some columns that may contain information that is not needed or would be unhelpful. There are also a couple columns with missing data but I will determine if that column is needed or not during my cleaning.

## Data Cleaning

In [90]:
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

I first want to find out the size of the data, so I will call the shape of the dataframe to determine how many rows and columns there are.

In [91]:
df_sales.shape

(8523, 12)

There are 8,523 rows and 12 columns. Next I want to find out the datatype of the values in each column.

In [92]:
df_sales.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

It looks there are integers, floats, and objects contained within the columns of the data set.

I'm now going to check the dataset for any duplicated rows. If there are any, I will drop them from the dataset.

In [93]:
df_sales.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8518    False
8519    False
8520    False
8521    False
8522    False
Length: 8523, dtype: bool

Because the size of the dataset is too large to display all rows, I am still unsure if the dataset contains duplicates. I will display the sum of duplicates in order to truly determine if there are any.

In [94]:
df_sales.duplicated().sum()

0

It looks like there are no duplicated rows in the set! From here, I want to identify the columns that have missing data. Based on the information from df_sales.info(), we can see that there are 2 columns with missing values: **Item_Weight** and **Outlet_Size**.
- I want to confirm this by running the data and checking for all values that are missing. The dataframe will return boolean values in place of the values in the dataset.

In [95]:
df_sales.isna()

Item_Identifier  Item_Weight  Item_Fat_Content  Item_Visibility  \
0               False        False             False            False   
1               False        False             False            False   
2               False        False             False            False   
3               False        False             False            False   
4               False        False             False            False   
...               ...          ...               ...              ...   
8518            False        False             False            False   
8519            False        False             False            False   
8520            False        False             False            False   
8521            False        False             False            False   
8522            False        False             False            False   

      Item_Type  Item_MRP  Outlet_Identifier  Outlet_Establishment_Year  \
0         False     False              False                      False   
1         False     False              False                      False   
2         False     False              False                      False   
3         False     False              False                      False   
4         False     False              False                      False   
...         ...       ...                ...                        ...   
8518      False     False              False                      False   
8519      False     False              False                      False   
8520      False     False              False                      False   
8521      False     False              False                      False   
8522      False     False              False                      False   

      Outlet_Size  Outlet_Location_Type  Outlet_Type  Item_Outlet_Sales  
0           False                 False        False              False  
1           False                 False        False              False  
2           False                 False        False              False  
3            True                 False        False              False  
4           False                 False        False              False  
...           ...                   ...          ...                ...  
8518        False                 False        False              False  
8519         True                 False        False              False  
8520        False                 False        False              False  
8521        False                 False        False              False  
8522        False                 False        False              False  

[8523 rows x 12 columns]

In [96]:
df_sales.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

There are 1463 missing values in **Item_Weight** and 2410 missing values in **Outlet_Size**. We'll first address the **Item_Weight** column. The values in that column are floats, so I will take the average of the values in the column and impute the value for the missing values.

In [97]:
df_sales['Item_Weight'].fillna(value = df_sales['Item_Weight'].mean(), inplace = True)
df_sales.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Now I want to fill in the missing data for the **Outlet_Size** column. There are 3 different values that exist in the column: 'High', 'Medium', 'Small.' This set of values is catergorical, not numerical, so finding the mean would not be the logical choice. Instead, I will use mode in order to use the data that most appears in the column, as the likelihood of that value in that column is greater.

In [98]:
outlet_size_mode = df_sales['Outlet_Size'].mode()
df_sales['Outlet_Size'].fillna(outlet_size_mode, inplace=True)

In [99]:
df_sales.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

After attempting to replace the **Outlet_Size** with the mode of the data, it appears that there may be more than 3 category options. I will run a value_counts for the column in order to determine the names of the values.

In [100]:
df_sales['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

It still appears that using the mode for the values in the column does not fill in the missing data. Based on the value_counts, 'Medium' is the most frequent value used, so I will instead impute 'Medium' in for the missing values.

In [101]:
df_sales['Outlet_Size'].fillna('Medium', inplace=True)

In [102]:
df_sales.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

We can see that imputing the value directly allowed me to fill in the missing data values in the column.

### Deleting Unnecessary Columns

I want to now delete any columns in the dataset that are unnecessary or have values that are unique to each row, as these will not help when the data is run through machine learning.

In [103]:
df_sales.drop(columns = ['Item_Identifier', 'Item_Visibility', 'Outlet_Identifier', 'Outlet_Location_Type'], inplace=True)
df_sales.head()

Item_Weight Item_Fat_Content              Item_Type  Item_MRP  \
0         9.30          Low Fat                  Dairy  249.8092   
1         5.92          Regular            Soft Drinks   48.2692   
2        17.50          Low Fat                   Meat  141.6180   
3        19.20          Regular  Fruits and Vegetables  182.0950   
4         8.93          Low Fat              Household   53.8614   

   Outlet_Establishment_Year Outlet_Size        Outlet_Type  Item_Outlet_Sales  
0                       1999      Medium  Supermarket Type1          3735.1380  
1                       2009      Medium  Supermarket Type2           443.4228  
2                       1999      Medium  Supermarket Type1          2097.2700  
3                       1998      Medium      Grocery Store           732.3800  
4                       1987        High  Supermarket Type1           994.7052

**Item_Identifier** contained unique values to each row, so they were deleted. I also deleted **Outlet_Identifier** for the same reason, and although it was present in more than one row, it was not prevalent enough to keep it. **Outlet_Location_Type** did not seem necessary as we already have information on the types of outlets.  And for the final column **Item_Visibility**, the data did not seem identify anything of importance with the data, and the values of each were very different from one another, so it was deleted as well.

Next I want to sort throught the columns with catergorical values in order to determine if there are any mistakes or values that have been named differently from the expected value.

In [83]:
df_sales['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Since there are inconsistencies, I will replace each value with either 'Low Fat' or 'Regular.' From the value counts, I can assume that 'LF' and 'low fat' can be converted to 'Low Fat', while 'reg' can be converted to 'Regular'.

In [84]:
df_sales.replace(['LF', 'low fat'], ['Low Fat', 'Low Fat'], inplace=True)
df_sales.replace(['reg'], ['Regular'], inplace=True)
df_sales['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

I will now check **Item_Type**.

In [85]:
df_sales['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

It appears that all the values in 'Item_Type' seem consistent. I now will check **Outlet_Type**.

In [86]:
df_sales['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

The values in the categorical columns have been checked for consistency, so the dataframe appears set. Lastly, I want to run a statistics summary for the numerical columns (mean, median, mode, etc.).

In [104]:
df_sales.head()

Item_Weight Item_Fat_Content              Item_Type  Item_MRP  \
0         9.30          Low Fat                  Dairy  249.8092   
1         5.92          Regular            Soft Drinks   48.2692   
2        17.50          Low Fat                   Meat  141.6180   
3        19.20          Regular  Fruits and Vegetables  182.0950   
4         8.93          Low Fat              Household   53.8614   

   Outlet_Establishment_Year Outlet_Size        Outlet_Type  Item_Outlet_Sales  
0                       1999      Medium  Supermarket Type1          3735.1380  
1                       2009      Medium  Supermarket Type2           443.4228  
2                       1999      Medium  Supermarket Type1          2097.2700  
3                       1998      Medium      Grocery Store           732.3800  
4                       1987        High  Supermarket Type1           994.7052

In [109]:
df_sales[{'Item_Weight', 'Item_MRP', 'Item_Outlet_Sales'}].describe().round(2)

Item_Outlet_Sales  Item_Weight  Item_MRP
count            8523.00      8523.00   8523.00
mean             2181.29        12.86    140.99
std              1706.50         4.23     62.28
min                33.29         4.56     31.29
25%               834.25         9.31     93.83
50%              1794.33        12.86    143.01
75%              3101.30        16.00    185.64
max             13086.96        21.35    266.89

I rounded the values to 2 decimal places in order to clean up the numerical values. I also did not include the year in the statistics summary as they act more as categories than they do numerically.

## Exploratory Visuals

## Explanatory Visuals